Upgrade storybook to v9

manzoorwanijk opened the pull request:
This came out as a part of #74135, inspired by #74309, but I thought to do it separately as an extra benefit for the future.

## What?

Upgrades Storybook from v8 to v9 <s>and converts the `storybook` directory into an npm workspace package to isolate its dependencies</s>. Updating to v10 is going to be a major task which may need some configuration changes to the whole monorepo, as [Storybook v10](https://storybook.js.org/docs/releases/migration-guide) is ESM only and doesn't work with `"moduleResolution": "node"`, changing that is going to be chaotic and thus it needs more thought.

https://github.com/WordPress/gutenberg/blob/0c4de409d7d5fabfb5e11004170eafea7e241599/tsconfig.base.json#L27

## Why?

- Storybook v9 brings performance improvements and updated APIs

## How?

- Upgraded Storybook and the related dependencies to v9
- Removed obsolete `@ts-expect-error` directives from story files
- Replaced the incompatible source-link and badge addons with compatible ones given on the storybook docs
- Fix a race condition in `npm run storybook:dev`. See https://github.com/WordPress/gutenberg/pull/74143#discussion_r2652925970

## Testing Instructions

1. Run `npm install`
2. Run `npm run clean:package-types && npm run build` — should complete without errors
3. Run `npm run storybook:build` — should build successfully
4. Run `npx http-server storybook/build` to server the built storybook files.
5. Go to the URL the server is available on.
6. Confirm that the stories load fine.
7. Run `npm run storybook:dev` — storybook should start at http://localhost:50240
8. Confirm that the badges and source links work fine

<img width="344" height="147" alt="Screenshot 2025-12-22 at 1 51 10 PM" src="https://github.com/user-attachments/assets/db01acd9-c939-498d-92ea-d12105f64a98" />
<img width="355" height="150" alt="Screenshot 2025-12-22 at 1 50 30 PM" src="https://github.com/user-attachments/assets/79cb95df-a703-404e-8366-ae4ad66cf01a" />

manzoorwanijk committed:
Upgrade Storybook from v8 to v9

manzoorwanijk committed:
Disable type-checking for node_modules

manzoorwanijk committed:
Move Storybook setup to use isolated workspace

This keeps our root package.json clean

manzoorwanijk committed:
Fix lints in package.json

manzoorwanijk committed:
Update lockfile

manzoorwanijk committed:
Fix incompatible storybook addons

SirLouen reviewed (commented):
Tested Windows 11 WSL2 with Ubuntu 24.04 and got the following errors:

<details>
<summary>Errors</summary>

```
✘ [ERROR] Could not resolve "@storybook/manager-api"

    ../node_modules/storybook-source-link/dist/esm/preset/manager.js:1:30:
      1 │ import { addons, types } from "@storybook/manager-api";
        ╵                               ~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "@storybook/manager-api" as external to exclude it from the bundle, which
  will remove this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "@storybook/manager-api"

    ../node_modules/@geometricpanda/storybook-addon-badges/dist/manager.mjs:1:47:
      1 │ import { addons, types, useStorybookApi } from '@storybook/manager-api';
        ╵                                                ~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "@storybook/manager-api" as external to exclude it from the bundle, which
  will remove this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "@storybook/components"

    ../node_modules/@geometricpanda/storybook-addon-badges/dist/manager.mjs:3:55:
      3 │ import { Separator, TooltipMessage, WithTooltip } from '@storybook/components';
        ╵                                                        ~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "@storybook/components" as external to exclude it from the bundle, which
  will remove this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "@storybook/components"

    ../node_modules/storybook-source-link/dist/esm/Tool.js:2:56:
      2 │ import { IconButton, TooltipMessage, WithTooltip } from "@storybook/components";
        ╵                                                         ~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "@storybook/components" as external to exclude it from the bundle, which
  will remove this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "@storybook/manager-api"

    ../node_modules/storybook-source-link/dist/esm/Tool.js:5:29:
      5 │ import { useParameter } from '@storybook/manager-api';
        ╵                              ~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "@storybook/manager-api" as external to exclude it from the bundle, which
  will remove this error and leave the unresolved path in the bundle.

Error: Build failed with 5 errors:
../node_modules/@geometricpanda/storybook-addon-badges/dist/manager.mjs:1:47: ERROR: Could not resolve "@storybook/manager-api"
../node_modules/@geometricpanda/storybook-addon-badges/dist/manager.mjs:3:55: ERROR: Could not resolve "@storybook/components"
../node_modules/storybook-source-link/dist/esm/Tool.js:2:56: ERROR: Could not resolve "@storybook/components"
../node_modules/storybook-source-link/dist/esm/Tool.js:5:29: ERROR: Could not resolve "@storybook/manager-api"
../node_modules/storybook-source-link/dist/esm/preset/manager.js:1:30: ERROR: Could not resolve "@storybook/manager-api"
    at failureErrorWithLog (/home/mcamargo/src/wordpress/wordpress-develop/src/wp-content/plugins/gutenberg/node_modules/esbuild/lib/main.js:1467:15)
    at /home/mcamargo/src/wordpress/wordpress-develop/src/wp-content/plugins/gutenberg/node_modules/esbuild/lib/main.js:926:25
    at runOnEndCallbacks (/home/mcamargo/src/wordpress/wordpress-develop/src/wp-content/plugins/gutenberg/node_modules/esbuild/lib/main.js:1307:45)
    at buildResponseToResult (/home/mcamargo/src/wordpress/wordpress-develop/src/wp-content/plugins/gutenberg/node_modules/esbuild/lib/main.js:924:7)
    at /home/mcamargo/src/wordpress/wordpress-develop/src/wp-content/plugins/gutenberg/node_modules/esbuild/lib/main.js:951:16
    at responseCallbacks.<computed> (/home/mcamargo/src/wordpress/wordpress-develop/src/wp-content/plugins/gutenberg/node_modules/esbuild/lib/main.js:603:9)
    at handleIncomingPacket (/home/mcamargo/src/wordpress/wordpress-develop/src/wp-content/plugins/gutenberg/node_modules/esbuild/lib/main.js:658:12)
    at Socket.readFromStdout (/home/mcamargo/src/wordpress/wordpress-develop/src/wp-content/plugins/gutenberg/node_modules/esbuild/lib/main.js:581:7)
    at Socket.emit (node:events:524:28)
    at addChunk (node:internal/streams/readable:561:12)
npm error Lifecycle script `storybook:build` failed with error:
npm error code 1
npm error path /home/mcamargo/src/wordpress/wordpress-develop/src/wp-content/plugins/gutenberg/storybook
npm error workspace @wordpress/storybook@1.0.0
npm error location /home/mcamargo/src/wordpress/wordpress-develop/src/wp-content/plugins/gutenberg/storybook
npm error command failed
npm error command sh -c storybook build -c . -o ./build
```

</details>

manzoorwanijk committed:
Replace the incompatible source-link and badge addons with compatible ones

manzoorwanijk committed:
Fix lints

manzoorwanijk committed:
Remove unused packages

manzoorwanijk commented on storybook/package.json:
```diff
@@ -0,0 +1,81 @@
+		"@storybook/manager-api": "9.0.0-alpha.1",
+		"@storybook/react-webpack5": "9.1.17",
+		"@storybook/source-loader": "8.4.7",
+		"@storybook/theming": "9.0.0-alpha.1",
```
Cleaned up

manzoorwanijk commented on storybook/package.json:
```diff
@@ -0,0 +1,81 @@
+		"@storybook/icons": "1.4.0",
+		"@storybook/manager-api": "9.0.0-alpha.1",
+		"@storybook/react-webpack5": "9.1.17",
+		"@storybook/source-loader": "8.4.7",
```
Cleaned up

SirLouen commented on .eslintrc.js:
```diff
@@ -430,6 +430,9 @@ module.exports = {
 				// Useful to add story descriptions via JSDoc without specifying params,
 				// or in TypeScript files where params are likely already documented outside of the JSDoc.
 				'jsdoc/require-param': 'off',
+				// Disable this rule as eslint-plugin-storybook 0.9.0 doesn't recognize the new
+				// `storybook/test` import path introduced in Storybook 9.
+				'storybook/use-storybook-expect': 'off',
```
Isn't it a good opportunity to update this package? It seems it's extremely old by now.

manzoorwanijk replied:
Its latest version works fine with the latest storybook version.

SirLouen replied:
What are the implications of upgrading it? Couldn't we add a new PR for this?

manzoorwanijk replied:
I mentioned about v10 in the PR description.

SirLouen replied:
I mean `eslint-plugin-storybook` not Storybook itself.

manzoorwanijk replied:
Its latest version doesn’t work with v9 of storybook.

SirLouen replied:
Not the latest version. We are currently in `0.9`, so maybe we could jump into `^9.0`

SirLouen replied:
Or maybe even `0.11.4`
https://github.com/storybookjs/eslint-plugin-storybook/releases/tag/v0.11.4

manzoorwanijk replied:
I tried `^9` but it caused the storybook build failure which turned out to be a rabbit hole to fix.

SirLouen replied:
Have you tried with 0.11.4?

SirLouen replied:
Anyway, it could be interesting to open a new PR for this purpose. Doesn't make any sense to keep upgrading the Storyblock and leave behind eslint.

manzoorwanijk replied:
`0.11.4` doesn't work either. Unfortunately, versions `0.10.0` and above all use `@typescript-eslint/utils@^8.x` which is incompatible with ESLint `8.3.0`. The last compatible version is `0.9.0`. The issue is that `eslint-plugin-storybook` jumped from using `@typescript-eslint/utils@^5.62.0` (in `0.9.0`) to `@typescript-eslint/utils@^8.8.1` (in `0.10.0+`), and the `v8` of `typescript-eslint/utils` has breaking changes with ESLint's flat config.

SirLouen replied:
Ideally, this should not be left behind. I'm going to take a look into `eslint-plugin-storybook` at least to do a step and see what effort is needed to upgrade to 0.11.4

SirLouen reviewed (commented):
Its building right for me now ✅

SirLouen commented on .eslintrc.js:
```diff
@@ -430,6 +430,9 @@ module.exports = {
 				// Useful to add story descriptions via JSDoc without specifying params,
 				// or in TypeScript files where params are likely already documented outside of the JSDoc.
 				'jsdoc/require-param': 'off',
+				// Disable this rule as eslint-plugin-storybook 0.9.0 doesn't recognize the new
+				// `storybook/test` import path introduced in Storybook 9.
```
I would update these comments to suggest removal once updates have taken place (and still it's a poor idea I wonder if anyone will remember to read this and fix it).

Ideally I would add a new issue/PR explaining the whole issue so its not left untracked.

manzoorwanijk replied:
I have added a TODO. I will create an issue for it once this PT is merged. Thanks.

SirLouen reviewed (changes requested):
I've been doing some research.

To update to anything >0.11.4 for `eslint-plugin-storybook` first we need ESLint v9 because the new versions refuse to use the old `.eslintrc.js` format, 

It seems this is being worked in #65648 (but it has gone a little bit stale, not going to lie). Still even after the implementation of ESLint 9, it doesn't assure, that `eslint-plugin-storybook` will be immediately updated hence, removed the line below.

Maybe a new Issue/Task to track everything in one place would be ideal.

manzoorwanijk committed:
Merge branch 'trunk' into upgrade/storybook

manzoorwanijk committed:
Add TODO for storybook eslint rule override

Added a TODO comment to remove the 'storybook/use-storybook-expect' rule override once eslint-plugin-storybook is upgraded to version 0.10.0 or higher.

manzoorwanijk committed:
We can still import from @storybook/react when declared as a dependency

This reduces the number of changes in this branch

SirLouen reviewed (approved):
Looking good 👍 
Let's hope that ESLint 9 lands anytime soon

mirka commented on storybook/preview.jsx:
```diff
@@ -168,7 +165,30 @@ export const parameters = {
+		links: {
+			// Disable default VSCode links
+			'component-vscode': () => undefined,
+			'story-vscode': () => undefined,
```
Let's also disable the `addon-powered-by`.

manzoorwanijk replied:
Done. Thanks

mirka commented on storybook/manager.js:
```diff
@@ -1,15 +1,18 @@
 
 addons.setConfig( {
 	sidebar,
+	tagBadges: [ ...tagBadges, ...defaultConfig ],
```
I don't think we'll need the default config.

manzoorwanijk replied:
Updated.

mirka commented on storybook/manager.js:
```diff
@@ -1,15 +1,18 @@
  * Internal dependencies
  */
-import theme from './theme';
+import { tagBadges } from './badges';
```
I think it'd be cleaner to programmatically map the canonical badge data directly in this manager file, rather than prepare a static export specifically for this.

```js
Object.entries( badges ).map(
	( [ key, { title, styles, tooltip } ] ) => ( {
		tags: `status-${ key }`,
		badge: {
			text: title,
			...( styles && { style: styles } ),
			tooltip,
		},
	} )
);
```

SirLouen replied:
Great catch. Still, this improvement could be left in `badges.js` given that it is more contextualized and keeps the `manager.js` focused on the configuration assembly.

manzoorwanijk replied:
This is cleaner. I updated it. Thanks

mirka commented on tsconfig.base.json:
```diff
@@ -14,7 +14,7 @@
 		"isolatedModules": true,
 
-		"skipDefaultLibCheck": true,
+		"skipLibCheck": true,
```
This might be something that should be updated in a dedicated PR?

SirLouen replied:
I also was a little bit doubtful about this, but long story short:

```
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
```

Probably this is not a strict cause of this implementation but a chain of circumstances.

manzoorwanijk replied:
`skipDefaultLibCheck` is deprecated, and it's now recommended to use `skipLibCheck`

[Docs](https://www.typescriptlang.org/tsconfig/#skipDefaultLibCheck)

Not making this change breaks the TS build in this branch due to Storybook now indirectly depending up `ast-types` which conflicts with certain types in the repo.

manzoorwanijk replied:
Created a separate PR for this change - #74239

mirka commented on storybook/package.json:
Was this a requirement for the Storybook upgrade? I feel like that's a separate conversation, given how the monorepo is currently set up to favor hoisted dev deps.

A Storybook upgrade PR is going to be quite big to begin with, so it will be easier to keep the scope as tight as possible.

SirLouen replied:
This is clearly the first thing that stands out when reviewing this PR.

But given that this is coming from #74135, which happens to be a proposal to implement `pnpm`, which happens to be the ultimate hoisting build tool, I understood that this was part of a bigger plan.

manzoorwanijk replied:
It's not a requirement for the upgrade per se. The reason for making it a workspace in the same PR was to ensure it was in the correct state needed for pnpm upgrade. As mentioned in the PR description, it helps us keep the root package.json cleaner; otherwise, with newer Storybook versions, including v9, you need a lot more dependencies in the root, which creates a lot of noise. So it's better to have it as a workspace to separate the concerns, just like #74212.

This PR is more about making it a workspace than upgrading the version.

ciampo replied:
I agree with the sentiment.

If we think that having a separate Storybook workspace would be beneficial, perhaps it would be better to do that as a preliminary step, and tackle the dependency update in a follow-up PR?

(nevertheless, thank you @manzoorwanijk for working on this — massively appreciated 🙏 )

manzoorwanijk replied:
Thanks. I made the dependency update a part of it as a confidence measure that the change is future-proof and won't break with the newer Storybook versions. I am happy to move out the version update, but I think it may be fine to do it together in this PR.

manzoorwanijk replied:
Alright, I have reverted the workspace related changes and kept this PR only for version upgrade.

mirka reviewed (commented):
Thanks for working on this! 🙏

I noticed a few more issues:

- The stories in Design System ▸ Components aren't loading anymore. Possibly an issue with the Theme toolbar item that only runs on these stories?
- There is also a [Storybook instance](https://github.com/WordPress/gutenberg/tree/trunk/test/storybook-playwright) for manually-triggered local playwright testing, which needs to be updated as well.

manzoorwanijk committed:
Disable all default links from source link addon

manzoorwanijk committed:
Fix hook dependency

manzoorwanijk committed:
Local addons cannot use JSX in newer Storybook versions

Replaces JSX with React.createElement and Fragment

manzoorwanijk committed:
Use badge data directly in manager

manzoorwanijk commented:
> * The stories in Design System ▸ Components aren't loading anymore. Possibly an issue with the Theme toolbar item that only runs on these stories?

Thank you for catching that. Newer storybook versions (from v9) expect the addons to be pre-built, which means that JSX inside them is not supported anymore. So, I updated the addon in `addons/design-system-theme` to `createElement` for now.

manzoorwanijk committed:
Remove @storybook/addon-toolbars from storybook instance

@storybook/addon-toolbars is now a part of Storybook core

manzoorwanijk commented:
> * There is also a [Storybook instance](https://github.com/WordPress/gutenberg/tree/trunk/test/storybook-playwright) for manually-triggered local playwright testing, which needs to be updated as well.

Updated. Thanks

Following https://github.com/storybookjs/storybook/pull/30871, `@storybook/addon-toolbars` was embedded into Storybook core in v9. We just needed to remove it from the addons list.

manzoorwanijk committed:
Merge branch 'trunk' into upgrade/storybook

manzoorwanijk committed:
Update package-lock.json

manzoorwanijk committed:
Merge branch 'trunk' into upgrade/storybook

manzoorwanijk committed:
Merge branch 'trunk' into upgrade/storybook

manzoorwanijk committed:
Revert the workspace conversion changes

manzoorwanijk committed:
Fix storybook:dev race condition with dev script

manzoorwanijk commented on bin/dev.mjs:
```diff
@@ -181,11 +178,30 @@ async function dev() {
+		} );
+
+		// Now signal that the build is ready
+		readyMarkerFile.create();
```
This fixes race condition in `npm run storybook:dev`. The `.dev-ready` marker was created before `wp-build --watch` finished its initial transpilation, allowing Storybook to start while packages were still being built. This latent bug became apparent with Storybook 9's faster startup, causing "Module not found" errors for `@wordpress/*` packages.

The fix waits for wp-build to output "Watching for changes..." before creating the marker.

CC: @youknowriad @jsnajdr

manzoorwanijk committed:
Merge branch 'trunk' into upgrade/storybook

manzoorwanijk committed:
Fix lints after merge conflicts

aduth commented:
We'll want to rebase this branch with #74190 having been merged, which added a few more `@ts-expect-error` I expect will need to be removed now ([see related comment](https://github.com/WordPress/gutenberg/pull/74190/changes#r2651764303)).

aduth commented on bin/dev.mjs:
```diff
@@ -181,11 +178,30 @@ async function dev() {
+				const output = data.toString();
+				process.stdout.write( output );
+				if ( output.includes( 'Watching for changes' ) ) {
+					resolve();
```
Should we detach the `stdout` event handler at this point? 

I also wonder if we should be `await`-ing this in the main path, or if it should allow the rest of the function to continue handling the clean-up. The worry being that if someone terminates the process before watching for files, none of that clean-up will happen.

i.e. Maybe it'd look more like...

```ts
new Promise( ( resolve ) => {
  // ...
} ).then( () => readyMarkerFile.create() );
```

manzoorwanijk replied:
That is a good idea. Thanks

manzoorwanijk replied:
Updated in https://github.com/WordPress/gutenberg/pull/74143/commits/116eb7b6e4495656296f81aedc92aa65a5f95b06

manzoorwanijk committed:
Merge branch 'trunk' into upgrade/storybook

manzoorwanijk committed:
Remove unused ts-expect-error comments

manzoorwanijk committed:
Detach stdout event handler and allow clean up on early termination

mirka reviewed (approved):
Major props for working on this, thank you so much! Things are looking good on my end 👍

manzoorwanijk committed:
Revert changes to bin/dev.mjs
